Cleanup the SIMD module

tristan957 opened the pull request:
- Add typing
- Port to typed_pos_args/typed_kwargs
- Clean up code
- Add to mypy

eli-schwartz reviewed (commented):
Only looked at the first two commits... the third commit seems to be changing some actual logic. Quick observation though.

tristan957 commented on mesonbuild/modules/simd.py:
```diff
@@ -15,58 +15,78 @@
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[SOURCES_KW.evolve(name=iset) for iset in ISETS],
+                  *STATIC_LIB_KWS,
```
Good idea. I'll remove it on my next push.

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -75,39 +75,52 @@ def __init__(self, interpreter: Interpreter):
         for iset in ISETS:
-            if iset not in kwargs:
+            del static_lib_kwargs[iset]
+        static_lib_kwargs = T.cast('kwtypes.StaticLibrary', static_lib_kwargs) # type: ignore
```
I think this could just be written as:
```python
local_keys = set(ISETS) + {'compiler'}
static_lib_kwargs = T.cast('kwtypes.StaticLibrary', {k: v for k, v in kwargs.items() if k not in local_keys))
```
which avoids the full copy with deletions, as well as the need to recast

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -75,39 +75,52 @@ def __init__(self, interpreter: Interpreter):
+        static_lib_kwargs = T.cast('kwtypes.StaticLibrary', static_lib_kwargs) # type: ignore
+
+        for iset in ISETS:
+            sources = kwargs.get(iset)
```
We shouldn't need to use get here, typed_kwargs should ensure this for us

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -84,5 +91,5 @@ def check(self, state, args, kwargs):
             result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
         return [result, conf]
 
-def initialize(*args, **kwargs):
+def initialize(*args: T.Any, **kwargs: T.Any) -> SimdModule:
     return SimdModule(*args, **kwargs)
```
```python
def initialize(interp: Interpreter) -> SimdModule:
    return SimdModule(Interpreter)
```

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,110 @@
+
+            my_name = f'{prefix}_{iset}'
+
+            my_kwargs = static_lib_kwargs.copy()
```
You need to somehow get this to `kwargs.StaticLibrary`. I think you could inherit that instead of `TypedDict` for `CheckKwargs`, or you could cast here.

dcbaker replied:
I think that inheriting will make things easier long term

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,110 @@
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[SOURCES_KW.evolve(name=iset) for iset in ISETS],
```
Thinking about this more (because of the mypy issue), I think you should import the `STATIC_LIB_KWS` kwargs  as well and add something like:
```python
*[a for a in STATIC_LIB_KWS if a.name != 'sources'],
```
Which will mean that SIMD will automatically get any new values added to the static lib module for free, and that you could drop the `'if 'sources' in kwargs:` check below.

dcbaker replied:
Well, actually you can't drop the `if 'source'` until the `allow_unknown` is dropped...

dcbaker reviewed (changes requested):
sorry, I know I approved and now I'm requesting changes, but I saw the mypy error and realized that there was a bit more we could improve here.

dcbaker reviewed (approved):
Looks great, thanks!

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -74,39 +74,46 @@ def __init__(self, interpreter: Interpreter):
-            args = compiler.get_instruction_set_args(iset)
-            if args is None:
-                mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
+
+            if not compiler.has_multi_arguments(cargs, state.environment)[0]:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
-            if args:
-                if not compiler.has_multi_arguments(args, state.environment)[0]:
-                    mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
-                    continue
-            mlog.log('Compiler supports %s:' % iset, mlog.green('YES'))
-            conf.values['HAVE_' + iset.upper()] = ('1', 'Compiler supports %s.' % iset)
-            libname = prefix + '_' + iset
-            lib_kwargs = {'sources': iset_fname,
-                          }
-            lib_kwargs.update(basic_kwargs)
-            langarg_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key)
-            all_lang_args = old_lang_args + args
-            lib_kwargs[langarg_key] = all_lang_args
-            result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
+
+            mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
+            conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
+
+            my_name = f'{prefix}_{iset}'
+            my_kwargs = static_lib_kwargs.copy()
+            my_kwargs['sources'] = sources # type: ignore
+
+            # Add compile args we derived above to those the user provided us
+            lang_args_key = compiler.get_language() + '_args'
```
As previously mentioned at https://github.com/mesonbuild/meson/pull/12104#discussion_r1290489180

> In general, this commit seems to have leaned very heavily into avoiding cases where git can recognize previous content in the diff.


Some of the decisions in this "clean up the code" confuse me quite a bit. As an example, why rename `langarg_key` to `lang_args_key`? Are the new variable names better? Keeping the old ones should allow git to show the actual semantic changes.

tristan957 replied:
Well it is documented as `<lang>_args`, not `<lang>_arg` (`c_args`, not `c_arg`

tristan957 replied:
Plus it is `old_lang_args`, not `old_langarg`. If anything, this particular rename is more consistent

tristan957 replied:
Anyway, removed the `my` prefix in favor of `lib`

eli-schwartz replied:
```diff
-libname = prefix + '_' + iset
+my_name = f'{prefix}_{iset}'
```
This name change feels very... Perl Programming Language.

But also, I don't actually like f-strings in such a case -- it actively feels harder to read when the f-string is almost entirely brackets denoting variables.

tristan957 replied:
No idea how that is more readable, but ok, here you go in the next push

eli-schwartz replied:
> Anyway, removed the `my` prefix in favor of `lib`

This is not what I asked for. :(

```diff
-libname = prefix + '_' + iset
+lib_name = prefix + '_' + iset
```

> If anything, this particular rename is more consistent

It's such an incredibly subjective "rename is more consistent", and that's the root of my objection. I'm not a fan of "submit a PR that contains blocks of 20 or 30 lines that are just one big hunk of removed followed by one big hunk of added, because a bunch of variables got renamed".

And adding an underscore to the variable name fails at my target objective, which was to make that line show up as unchanged and divide up the `git log -p` view.

eli-schwartz replied:
> Well it is documented as `<lang>_args`, not `<lang>_arg` (`c_args`, not `c_arg`

I am not sure why the internal python code names need to match the precise spelling, underscore style, and pluralization of the meson.build form. It is simply not important! But being able to `git diff` is important, and that means that bikesheddy variable renames are a bad thing when they mean that the entire file is a solid wall of red and green.

eli-schwartz replied:
```diff
diff --git a/mesonbuild/modules/simd.py b/mesonbuild/modules/simd.py
index 2be4e9164..0a5762b46 100644
--- a/mesonbuild/modules/simd.py
+++ b/mesonbuild/modules/simd.py
@@ -88,29 +88,29 @@ class SimdModule(ExtensionModule):
         for iset in ISETS:
             sources = kwargs[iset] # type: ignore
 
-            cargs = compiler.get_instruction_set_args(iset)
-            if cargs is None:
+            args = compiler.get_instruction_set_args(iset)
+            if args is None:
                 mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
 
-            if not compiler.has_multi_arguments(cargs, state.environment)[0]:
+            if not compiler.has_multi_arguments(args, state.environment)[0]:
                 mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
 
             mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
             conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
 
-            lib_name = prefix + '_' + iset
+            libname = prefix + '_' + iset
             lib_kwargs = static_lib_kwargs.copy()
             lib_kwargs['sources'] = sources # type: ignore
 
             # Add compile args we derived above to those the user provided us
-            lang_args_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, lang_args_key) # type: ignore
-            all_lang_args = old_lang_args + cargs
-            lib_kwargs[lang_args_key] = all_lang_args # type: ignore
+            langarg_key = compiler.get_language() + '_args'
+            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key) # type: ignore
+            all_lang_args = old_lang_args + args
+            lib_kwargs[langarg_key] = all_lang_args # type: ignore
 
-            lib = self.interpreter.build_target(state.current_node, (lib_name, []), lib_kwargs, build.StaticLibrary)
+            lib = self.interpreter.build_target(state.current_node, (libname, []), lib_kwargs, build.StaticLibrary)
 
             result.append(lib)
 
```

This makes it significantly easier to follow the changes that were made, the commit still achieves everything it seems to be trying to achieve, but without unnecessary variable renames it also becomes *incredibly* easier to actually... review.

As I pointed out in August.

dcbaker replied:
```diff
-libname = prefix + '_' + iset
+my_name = f'{prefix}_{iset}'
```

I'm sorry, what. This is *exactly* the case that f-strings make more readable.

eli-schwartz replied:
It's easier to read code when the syntax highlighting for variables is replaced by syntax highlighting for a string? Or it's easier to read code when you're reading strings that are mostly code, but with a single character of an actual string in the middle?

I think f-strings make things more readable when the majority of the string is, well, a string, and the context switch of scanning over to the right in order to find out what `{}` or `%s` is, is eliminated. Reducing context switches is fantastic! Requiring me to give up syntax highlighting isn't fantastic, but in simple cases it's a highly worthwhile tradeoff.

tristan957 replied:
This is what I see. I am curious now what you see in your theme.

![image](https://github.com/mesonbuild/meson/assets/11936772/1a018236-823a-41d9-a971-c0bbfbd56b35)

dcbaker replied:
Mine highlights inside the f-string as well.

tristan957 commented:
Seems like there is some disruption in Docker's service.

tristan957 commented:
Here is the objectively worse code as requested. Please look at this block of almost 30 lines with no spaces and tell me how that is more readable.

tristan957 commented:
> -            cargs = compiler.get_instruction_set_args(iset)
-            if cargs is None:
+            args = compiler.get_instruction_set_args(iset)
+            if args is None:

Why would I want to reassign to a parameter of the function? That is just asking for trouble.

tristan957 commented:
I would very much like to be done with this PR. It is seriously a waste of my time at this point.

tristan957 commented:
> ```diff
> -            cargs = compiler.get_instruction_set_args(iset)
> -            if cargs is None:
> +            args = compiler.get_instruction_set_args(iset)
> +            if args is None:
> ```
> 
> Why would I want to reassign to a parameter of the function? That is just asking for trouble.

If this is actually how you want the code, please raise an issue with mypy:

```
mesonbuild/modules/simd.py:87:20: error: Incompatible types in assignment (expression has type "list[str]", variable has type "tuple[str]")  [assignment]
mesonbuild/modules/simd.py:91:49: error: Argument 1 to "has_multi_arguments" of "Compiler" has incompatible type "tuple[str]"; expected "list[str]"  [arg-type]
mesonbuild/modules/simd.py:102:29: error: No overload variant of "__add__" of "list" matches argument type "tuple[str]"  [operator]
mesonbuild/modules/simd.py:102:29: note: Possible overload variants:
mesonbuild/modules/simd.py:102:29: note:     def __add__(self, list[Any], /) -> list[Any]
mesonbuild/modules/simd.py:102:29: note:     def [_S] __add__(self, list[_S], /) -> list[_S | Any]
```

tristan957 commented:
I don't understand how preserving 6 lines of original code within the for loop is some kind of win. Can you please explain @eli-schwartz?

tristan957 commented:
I will not be pushing any more to this branch until I get better guidance on how to make this PR mergeable given that the commit in question is trying to **CLEANUP** the code and **MAKE** it mypy compliant.

dcbaker commented:
Please fix the mypy naming issues so we're not overriding args. I'd put that in a separate commit, probably the first commit of the series. If you still want to move to f-strings, do that in a separate commit, at the end would be fine.

tristan957 commented:
I am all for the crusade of making git diffs as small as possible. I have already told many contributors that in the past to prep them for reviews, but my PR is basically touching every line of code in that function regardless of if I want to or not as you can see in the screenshot I posted above, so I just don't understand why we are trying to achieve this impossible standard of perfection unless we don't want to add mypy compliance or leave an unnecessary `if args:` after we already did a continue.

It isn't like the changes I had were gratuitous. I moved variable declarations next to each other, gave breathing room to related blocks of code, and renamed variables in such a way that I think the outcome is more readable. Sure, my initial `my_` prefix sucked, but `lib_` at least makes sense with the `lib` variable that already exists. `libname` really does stick out when eyeballing the code next to `lib_kwargs`. And I really don't get how `langarg_key` is a better variable name than `lang_args_key`. Why would we not want variable names to line up with documentation? That raises the barrier of contribution slightly, but for no good reason.

eli-schwartz commented:
> Please look at this block of almost 30 lines with no spaces and tell me how that is more readable.

At no point whatsoever did I suggest avoiding the addition of spaces is required or even good.

I did say that preserving baseline references for semantic diffability is good, and that *subjective* variable renames that serve no code purpose, like renaming `libname` to `lib_name`, is disruptive to that.

> Why would I want to reassign to a parameter of the function? That is just asking for trouble.

I agree that is a good reason to rename a variable. I didn't object to it in my comments, but I did happen to include it in the example diff, which I specifically cared about the second half, that provides significant context.

```diff
-            mlog.log('Compiler supports %s:' % iset, mlog.green('YES'))
-            conf.values['HAVE_' + iset.upper()] = ('1', 'Compiler supports %s.' % iset)
+
+            mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
+            conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
+
             libname = prefix + '_' + iset
-            lib_kwargs = {'sources': iset_fname,
-                          }
-            lib_kwargs.update(basic_kwargs)
+            lib_kwargs = static_lib_kwargs.copy()
+            lib_kwargs['sources'] = sources # type: ignore
+
+            # Add compile args we derived above to those the user provided us
             langarg_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key)
+            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key) # type: ignore
             all_lang_args = old_lang_args + args
-            lib_kwargs[langarg_key] = all_lang_args
-            result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
+            lib_kwargs[langarg_key] = all_lang_args # type: ignore
+
+            lib = self.interpreter.build_target(state.current_node, (libname, []), lib_kwargs, build.StaticLibrary)
+
+            result.append(lib)
```

> but my PR is basically touching every line of code in that function regardless of if I want to or not as you can see in the screenshot I posted above

?

tristan957 commented:
> ?

Look at your own diff that you just posted. Every line but 3 got touched. What do you mean `?`?

tristan957 commented:
> that subjective variable renames that serve no code purpose, like renaming libname to lib_name, is disruptive to that.

That is an _objectively_ better variable name. How can you even argue that as subjective?

eli-schwartz commented:
> Sure, my initial `my_` prefix sucked, but `lib_` at least makes sense with the `lib` variable that already exists. `libname` really does stick out when eyeballing the code next to `lib_kwargs`.

libname flows together very well, so well that we use it all over the codebase in contrast to lib_name. I am skeptical of calling it such a terrible name that it must be rewritten.

libkwargs reads very awkwardly IMHO, but it's all subjective either way, so.

When it's all subjective, why rewrite from one subjectivity to another?

> And I really don't get how `langarg_key` is a better variable name than `lang_args_key`. Why would we not want variable names to line up with documentation?

I don't see either one as being a better or worse variable name, what I see langarg_key as, is context-sensitive existing code that is being replaced by something that isn't validating the replacement with a compelling "this is better".

I do not at all buy the logic that the documentation has anything to do with it??

> so I just don't understand why we are trying to achieve this impossible standard of perfection

"impossible standard of perfection"

> I will not be pushing any more to this branch until I get better guidance on how to make this PR mergeable given that the commit in question is trying to **CLEANUP** the code and **MAKE** it mypy compliant.

On August 10, in the original PR this was split out from, I said that I found this code hard to review because it was a giant unbroken block of changed code. My perspective is that ever since then nothing changed, but you were happy to ***mock*** my concerns several times, then offer to make changes that didn't actually do what I asked.

eli-schwartz commented:
> Look at your own diff that you just posted. Every line but 3 got touched. What do you mean `?`?

I mean exactly what I have said a ***bunch*** of times already but you keep on dismissing. It provides anchor points in the diff -- instead of one giant block of rewritten code, it is four small blocks, each of which can be eyeballed separately.

tristan957 commented:
> you were happy to mock my concerns several times, then offer to make changes that didn't actually do what I asked.

That isn't mocking. That is my frustration of having basically wasted a whole work day because you wanted some anchor points for what was a diff of a few tens of lines, especially when it had already been approved by dcbaker and was one job away from being merged.

I wanted to merge this and then move onto the issue I mentioned on IRC, but instead I had to babysit this PR.

How would you feel spending a day for what you deem to be churn that you think makes the code worse than you had it?

tristan957 commented:
Anyway, I am not making any more changes to this, and I am done. Please feel free to push to my branch. I am done not having my time valued over a few anchor points.

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -78,35 +78,38 @@ def check(self, state: ModuleState, args: T.Tuple[str], kwargs: CheckKw) -> T.Li
-                mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
+
+            if not compiler.has_multi_arguments(compile_args, state.environment)[0]:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
-            if args:
-                if not compiler.has_multi_arguments(args, state.environment)[0]:
-                    mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
-                    continue
-            mlog.log('Compiler supports %s:' % iset, mlog.green('YES'))
-            conf.values['HAVE_' + iset.upper()] = ('1', 'Compiler supports %s.' % iset)
+            mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
+            conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
+
             libname = prefix + '_' + iset
-            lib_kwargs = {'sources': iset_fname,
-                          }
-            lib_kwargs.update(basic_kwargs)
+            lib_kwargs = static_lib_kwargs.copy()
+            lib_kwargs['sources'] = sources # type: ignore
+
+            # Add compile args we derived above to those the user provided us
             langarg_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key)
-            all_lang_args = old_lang_args + args
-            lib_kwargs[langarg_key] = all_lang_args
-            result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
+            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key) # type: ignore
+            all_lang_args = old_lang_args + compile_args
+            lib_kwargs[langarg_key] = all_lang_args # type: ignore
```
These type: ignores, can be removed if you instead add `sources: FileOrString` as attributes of `interpreter/kwargs.py -> _BuildTarget`.

tristan957 replied:
Does this require any other build typing changes?

eli-schwartz replied:
I made that change and mypy ran cleanly on the entire codebase afterward.

I think at the moment we are mostly relying on interpreter.py not actually getting checked, to hide the fact that most keys in kwargs.py aren't actually filled in. They tend to get filled in while someone is actually working on the corresponding type_checking.py updates, but it's not a requirement I don't think.

tristan957 replied:
I will add the sources kwarg in a separate commit. Thank you for pointing that out. I will use a `T.Optional[T.List[FileOrString]]`

tristan957 replied:
Should it be on _BaseBuildTarget

eli-schwartz replied:
Thinking again on this:
- the reason it works for simd.py is, more or less, that we are assigning one CheckKw.sources to another, so the type simply does not matter and mypy will be happy since it is the *same*
- you are of course right that this should be a T.List, duh, I do not know what I was thinking
- There's actually several different things it can be as well, in type_checking.py it is defined as: `T.List[T.Union[FileOrString, CustomTarget, CustomTargetIndex, GeneratedList]]` but this is the `declare_dependency(sources:)` definition -- we actually want to support StructuredSources and ExtractedObjects too
  - this is a bug in the PR, I *think*. People could pass ExtractedObjects to a simd.check() and it would currently work because static_library() supports it. Probably no one uses structured_sources here since that implies building a rust library...
- `_BaseBuildTarget` suggests that `jar()` accepts the kwarg. But jar is special since it accepts fewer types, namely, does not accept structured_sources IIRC.

dcbaker replied:
That's wrong in several different ways.

1. mypy now believes that `sources` is always present. Its not
2. the type is wrong, it's not `FileOrString`, its `SourceVaragsTypes`
3. this is going to be broken again anyway once the correct annotation is added because lists are invariant

This is tangential to this change. mypy will happily error on "unneeded type ignore", and we should take advantage of that

tristan957 replied:
So I should revert?

tristan957 replied:
Here is static library: https://mesonbuild.com/Reference-manual_functions.html#static_library.

eli-schwartz replied:
> That's wrong in several different ways.
> 
>  1. mypy now believes that `sources` is always present. Its not

So does this mean CheckKw is badly defined?

>  2. the type is wrong, it's not `T.List[FileOrString]`, its `SourceVaragsTypes`

Yeah, I pointed that out already. Sorry for the initial red herring. :(

>  3. this is going to be broken again anyway once the correct annotation is added because lists are invariant

If it breaks anything it will break it identically now or in the future, so either we *never* handle it correctly or we... cast it? I guess? We should be able to do it today, I think.

eli-schwartz replied:
Regardless of anything else, this conversation made me discover a regression: #12302

tristan957 replied:
> So does this mean CheckKw is badly defined?

Yes, whoops.

tristan957 replied:
Where is `SourceVaragsTypes`?

eli-schwartz replied:
That typo is better spelled `SourcesVarargsType` and is in type_checking.py

dcbaker replied:
Casting Is i think the only way to do that today.

> So does this mean CheckKw is badly defined?

technically yes. but in a not at all easy way to fix. It really should have `sources: NotRequired[SourceVarargsTypes]`, but I think the slightly wrong annotation is preferable to having to redefine everything and keep two copies or even the mad inheritance hacks.

Also, I really dislike the `sources` keyword argument, since it just duplicates positional arguments. This is just another reason to not like it.

tristan957 replied:
Ctrl + F returns nothing for me

tristan957 replied:
Guess I just needed a rebase...

tristan957 replied:
So do I keep my changes to `_BaseBuildTarget` (but switch to SourcesVarargsType), move those changes to `_BuildTarget`, or add the type ignores back and forget it

eli-schwartz replied:
This is now independently implemented so I think the remaining typing issue is to reuse `sse: SourcesVarargsType` etc. in `CheckKw`.

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -13,53 +13,77 @@
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[SOURCES_KW.evolve(name=iset) for iset in ISETS],
```
Re-raising this for visibility so we don't forget:

SOURCES_KW is intended and written for declare_dependency, and has the wrong types for library() and friends -- specifically it excludes a couple of things you can add to a library but not to a declared dependency. So I think it's wrong for simd.check as well, we need a new dedicated KwargInfo that aligns with what library() expects and does.

tristan957 replied:
Are those all the things that are in SourcesVarargsType?

tristan957 replied:
All this FeatureNew discovery is going to be a `****`

dcbaker replied:
https://github.com/mesonbuild/meson/pull/12303

dcbaker replied:
@tristan957 sources keyword landed. Hopefully with the rebase this will make your life easier

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,109 @@
-            raise mesonlib.MesonException('Argument must be a string.')
-        if 'compiler' not in kwargs:
-            raise mesonlib.MesonException('Must specify compiler keyword')
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[BT_SOURCES_KW.evolve(name=iset) for iset in ISETS],
+                  *[a for a in STATIC_LIB_KWS if a.name != 'sources'],
+                  allow_unknown=True) # Because we also accept STATIC_LIB_KWS, but build targets have not been completely ported to typed_pos_args/typed_kwargs.
```
I *think* we need to add `permitted_kwargs(build.known_stlib_kwargs)` here since we are removing the invocation of interpreter.func_static_lib and using interpreter.build_target directly.

Other than that, this looks good.

eli-schwartz replied:
Right, clearly I meant that even if I didn't actually know I meant that. :D
